### ISSUE TITLE
--strictNullChecks compiler options reports wrong error location

### ISSUE THREAD

**Issue by jmlopez-rod** [Microsoft/TypeScript#11659]
**TypeScript Version:**  2.0.3

**Code**

```ts
interface IExample {
  array?: any[];
}

const obj: IExample = {
  array: [],
};
obj.array.push(1);
```

**Expected behavior:**

Report error on line 8:3 since `array` is optional and could be undefined.

**Actual behavior:** 

```
8:1  Object is possibly 'undefined'.  TS2532
```

Instead it reports it at the beginning of `obj` causing me to believe that `obj` may possibly be undefined when the one that could be undefined is the `array` property.

**Comment by DanielRosenwasser** [comment]
The problem is that our error emit span doesn't tell you the length of the error. Try running with `--pretty`, or [use an editor with TypeScript support](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Editor-Support) to see the range properly.

Command line with `--pretty`:

![image](https://cloud.githubusercontent.com/assets/972891/19413498/da37153e-92e3-11e6-9a6a-285af83dd375.png)


VS Code:

![image](https://cloud.githubusercontent.com/assets/972891/19413500/e91dfbbc-92e3-11e6-8791-c26154c6c1af.png)

**Comment by jmlopez-rod** [comment]
Regardless of the pretty flag, (10,1) is not the correct location. Imagine having a very long nested statement:

```
obj.prop1.prop2.prop3.prop4
```

Knowing the line number doesn't really help us determine where the error is. What if the error was because prop3 was the cause? Also, I'm using the API to collect the error locations and format my own messages so I wouldn't know if the `pretty` flag would do me any good. I really think the error message should be a bit more helpful, at least mention what property is at fault if you are going to be pointing out the beginning of object. In the latest case it should probably say:

```
Property 'prop1.prop2.prop3' of 'obj' is possibly undefined.
```

**Comment by DanielRosenwasser** [comment]
> Also, I'm using the API to collect the error locations and format my own messages

If you're using the API, each `Diagnostic` should have a start and a length for a given `SourceFile`, so the standard output stream shouldn't matter for that purpose.

> In the latest case it should probably say:

> ```
'obj.prop1.prop2.prop3' is possibly undefined.
```

I definitely agree that we can do better. We're actually taking pull requests for #8274.

**Comment by jmlopez-rod** [comment]
Thank you for pointing out the other issue, it is exactly like this one. I hope the error message can get better.
